### DISCUSSION TITLE
Need Help with Writing my First Xposed Module to Bypass Audio Recording API Restrictions

### DISCUSSION THREAD

**Discussion by DavidBerdik** [LSPosed/LSPosed#1187] [Q&A] [answered]
Hello LSPosed Community,

Although I have been an LSPosed user for about a year now, I have never tried digging in to building an LSPosed module. I recently became interested in building, however with the goal of bypassing the Android API's restrictions on what apps can and cannot have their audio captured.

I started by looking at the Android Developer documentation for capturing video and audio playback at https://developer.android.com/guide/topics/media/av-capture, and from there, started digging through the Android source code to try identifying what functions need hooked in order to make this work. I decided that the first one I would try would be the function that I believe processes the `android:allowAudioPlaybackCapture` value in the `AndroidManifest.xml` file for an app. The function call that I believe to be responsible for this is available [here](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/content/pm/parsing/ParsingPackageImpl.java;l=2539).

Since I have no experience with building Xposed modules, I decided that instead of starting a new project from nothing, I would instead use the [DisableFlagSecure project code](https://github.com/LSPosed/DisableFlagSecure/) as a starting point and modify it to hook the function. Here is what I came up with as my modified function hook.

```java
package com.test.universal_audio_capture_enabler;

import android.media.AudioAttributes;
import android.os.Build;

import de.robv.android.xposed.IXposedHookLoadPackage;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XC_MethodReplacement;
import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.XposedHelpers;
import de.robv.android.xposed.callbacks.XC_LoadPackage;

public class AudioCaptureEnabler implements IXposedHookLoadPackage {
    // Useful Reference: https://developer.android.com/guide/topics/media/av-capture

    @Override
    public void handleLoadPackage(XC_LoadPackage.LoadPackageParam loadPackageParam) {
        // Filter by package names with "android" in them and limit hooking to Android 11.
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {

            // Bypass "allowAudioPlaybackCapture" in AndroidManifest.xml
            /* https://cs.android.com/android/platform/superproject/+/master:frameworks/
                base/core/java/android/content/pm/parsing/ParsingPackageImpl.java;l=2539 */
            try {
                Class<?> parsingPkgImpl = XposedHelpers.findClass(
                        "android.content.pm.parsing.ParsingPackageImpl",
                        loadPackageParam.classLoader);

                XposedHelpers.findAndHookMethod(
                        parsingPkgImpl,
                        "setAllowAudioPlaybackCapture",
                        XC_MethodReplacement.returnConstant(true));
            } catch (Throwable t) {
                XposedBridge.log(t);
            }
        }
    }
}
```

After making this change, compiling the app, installing it on my phone, enabling "System Framework" as well as another app to test audio recording with in the LSPosed manager app, and rebooting, I tried testing to see if the hook was working properly and instead had the following error in the logs:

```
[ 2021-09-27T20:45:35.647     1000:  1693:  1693 E/LSPosed-Bridge  ] java.lang.NoSuchMethodError: android.content.pm.parsing.ParsingPackageImpl#setAllowAudioPlaybackCapture()#exact
	at de.robv.android.xposed.XposedHelpers.findMethodExact(Unknown Source:79)
	at de.robv.android.xposed.XposedHelpers.findAndHookMethod(Unknown Source:27)
	at com.test.universal_audio_capture_enabler.AudioCaptureEnabler.handleLoadPackage(AudioCaptureEnabler.java:29)
	at de.robv.android.xposed.IXposedHookLoadPackage$Wrapper.handleLoadPackage(Unknown Source:2)
	at de.robv.android.xposed.callbacks.XC_LoadPackage.call(Unknown Source:6)
	at de.robv.android.xposed.callbacks.XCallback.callAll(Unknown Source:28)
	at K0.beforeHookedMethod(Unknown Source:39)
	at de.robv.android.xposed.LspHooker.handleHookedMethod(Unknown Source:75)
	at LspHooker_.startBootstrapServices(Unknown Source:11)
	at com.android.server.SystemServer.run(SystemServer.java:596)
	at com.android.server.SystemServer.main(SystemServer.java:414)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:925)
```

Does anyone have any idea what I am doing wrong? In an earlier version that I tried to compile, I had a conditional just like the one in the Disable FLAG_SECURE code hook that read `if (loadPackageParam.packageName.equals("android")) {`, but it seemed like the module was not doing anything with that in there, so I tried removing it and running again, and this was the result.

I ran in to this same issue whenever I was trying to hook two other functions for this project, but have left them out of here for now since I think one example is enough.

Apparently, I am also going to have to add a native hook for the C-based `AAudio` API, which I am almost certainly going to need help with as well, but before I do that, I would like to sort out the Java-based hooks first.

Thank you in advance!

**Comment by DavidBerdik** [comment, marked as answer]
As an update, I figured out how to fix the above error. It turns out that I forgot to include the function arguments as part of the hook. Adding that fixed the issue!

**Reply by w311ang** [reply]
Could you share this work?

**Reply by DavidBerdik** [reply]
@w311ang I've made the GitHub repo for this project public here: https://github.com/DavidBerdik/Universal-Audio-Capture-Enabler-for-Android

It's very incomplete and broken though, so I wouldn't trust it as a reliable resource. The code changes that I've made are all on `audio-capture-enabler-fork`. As you can see, I was trying to use the source code from DisableFlagSecure as a starting template to the get the project started. I've set this project aside in recent months to focus on other things, however, if you can help me or know of someone who can help me, I would greatly appreciate it.

**Reply by w311ang** [reply]
@DavidBerdik Thanks for your work. It cannot be used for applications that use telephony audio as the audio output, such as a video conferencing app.  Screen recorded video has no sound.  It may be necessary to redirect this audio to media audio to be able to be recorded.

**Reply by DavidBerdik** [reply]
@w311ang You're welcome! I already knew about  the limitations you described here, and that's exactly what I was hoping to resolve with this project. Unfortunately, I never managed to get it off the ground. Do you have the knowledge necessary to get it working? I'd like to revisit this project again in the future, but it's not a priority at the moment.

**Reply by w311ang** [reply]
@DavidBerdik From the information I can find, this problem can be easily solved by running restricted applications in the Android virtual machine on the phone, but doing so obviously consumes more unnecessary performance.  The virtual machine I tried to use is [VPhoneGaGa](http://www.gsxnj.cn/mobile/index/?language=en-us), which can record sound when I record the screen externally.

**Reply by DavidBerdik** [reply]
@w311ang It's certainly nice to know that this option exists and works for at least some use cases, but it's certainly not the ideal solution. Getting an Xposed module to bypass the restriction is definitely possible. It's just a question of how.

**Reply by w311ang** [reply]
@DavidBerdik Here I have a Magisk module [AudioHQ](https://alcatraz323.github.io/audiohq/index_en.html) that can adjust the volume of each application. It is implemented by modifying some Android libs, perhaps [its source code](https://github.com/Alcatraz323?tab=repositories&q=audiohq&type=&language=&sort=) to help with your work, or ask the author to collaborate with you.

**Reply by DavidBerdik** [reply]
@w311ang Interesting! Thank you for sharing. When time permits, I will do some exploration of this to see what I can turn up.
